### PR TITLE
Refactor lazymaps vaults in preparation for localized ids

### PR DESCRIPTION
**Pull request by talekhinezh**

### CODE REVIEW

File: radix-engine/src/engine/process.rs [hunk 1]
```diff
@@ -1046,56 +1093,90 @@ impl<'r, 'l, L: SubstateStore> Process<'r, 'l, L> {
...
 
         Ok(CreateLazyMapOutput { mid })
     }
 
+    fn get_local_lazy_map(
+        &mut self,
+        mid: Mid,
+    ) -> Result<(&mut LazyMap, LazyMapState), RuntimeError> {
+        match self.unclaimed_lazy_maps.get_mut(&mid) {
+            Some((map, _, _)) => Ok((map, ClaimedByLazyMap(mid))),
```
Comment (iamyulong) [inline radix-engine/src/engine/process.rs:1159]: It's a bit strange to consider an unclaimed map  as claimed by itself.

File: radix-engine/src/engine/process.rs [hunk 2]
```diff
@@ -67,12 +79,19 @@ pub struct Process<'r, 'l, L: SubstateStore> {
...
     moving_bucket_refs: HashMap<Rid, BucketRef>,
-    /// Lazy maps which haven't been assigned to a component or lazy map yet.
-    unclaimed_lazy_maps: HashMap<Mid, LazyMap>,
+
     /// Vaults which haven't been assigned to a component or lazy map yet.
     unclaimed_vaults: HashMap<Vid, Vault>,
+    /// Vaults which have been assigned to a lazy map but not a component
+    claimed_vaults: HashMap<Vid, (Vault, Mid)>,
+    /// Lazy maps which haven't been assigned to a component or lazy map yet.
+    unclaimed_lazy_maps: HashMap<Mid, (LazyMap, HashSet<Mid>, HashSet<Vid>)>,
```
Comment (iamyulong) [inline radix-engine/src/engine/process.rs:88]: Suggest to have a structure with named fields for the value tuple. To some extend, we should probably have a forest representation.

File: radix-engine/src/engine/process.rs [hunk 3]
```diff
@@ -67,12 +79,19 @@ pub struct Process<'r, 'l, L: SubstateStore> {
...
+    /// Lazy maps which have been assigned to a lazy map but not a component
+    claimed_lazy_maps: HashMap<Mid, (LazyMap, Mid)>,
+
     /// Components which have been loaded and possibly updated in the lifetime of this process.
-    updating_components: HashMap<Address, ValidatedData>,
+    component_state: ComponentState,
```
Comment (iamyulong) [inline radix-engine/src/engine/process.rs:93]: Nice simplifier!

File: radix-engine/src/engine/process.rs [hunk 4]
```diff
@@ -1046,56 +1093,90 @@ impl<'r, 'l, L: SubstateStore> Process<'r, 'l, L> {
...
                     .unclaimed_lazy_maps
                     .remove(&mid)
                     .ok_or(RuntimeError::LazyMapNotFound(mid))?;
-                self.track.put_lazy_map(mid, lazy_map);
+                self.track.put_lazy_map(component_address, mid, lazy_map);
+                for descendent_mid in mids {
+                    let (descendent_lazy_map, _) =
+                        self.claimed_lazy_maps.remove(&descendent_mid).unwrap();
+                    self.track
+                        .put_lazy_map(component_address, descendent_mid, descendent_lazy_map);
```
Comment (iamyulong) [inline radix-engine/src/engine/process.rs:1116]: How about the descendent of descendent?

**Review by iamyulong — COMMENTED** [review]
I have to admit that I didn't understand fully the logic, but would like to push some early comments I had in mind.

File: radix-engine/src/engine/process.rs [hunk 5]
```diff
@@ -49,6 +50,26 @@ macro_rules! re_warn {
...
+    Empty,
+    Loaded(ValidatedData),
+    Saved,
+}
+
+#[derive(Debug)]
+struct UnclaimedLazyMap {
+    lazy_map: LazyMap,
+    child_lazy_maps: HashMap<Mid, LazyMap>,
+    child_vaults: HashMap<Vid, Vault>,
```
Comment (iamyulong) [inline radix-engine/src/engine/process.rs:68]: Worth commenting that these are not only direct children, but all descendants.

**Review by iamyulong — COMMENTED** [review]
I think there might be a flaw in the tree implementation (the LI about all descendants is broken), but could not verify it. Please find the code below that failed:
```rust
use scrypto::prelude::*;

blueprint! {
    struct Hello {
        maps: LazyMap<u32, LazyMap<u32, LazyMap<u32, LazyMap<String, String>>>>
    }

    impl Hello {
        pub fn new() -> Component {
            let map0 = LazyMap::new();
            let map1 = LazyMap::new();
            let map2 = LazyMap::new();
            let map3 = LazyMap::new();
            let map4 = LazyMap::new();
            map2.insert(3u32, map3);
            map0.insert(1u32, map1);
            map0.insert(2u32, map2);

            // I'm trying to verify the behavior of inserting a node to the leaf 
            // of a tree of maps, but got a LazyMapNotFound error.
            let map2 = map0.get(&2u32).unwrap();
            let map3 = map2.get(&3u32).unwrap();
            map3.insert(4u32, map4);
            
            Hello {
                maps: map0
            }.instantiate()
        }
    }
}
```

File: radix-engine/src/engine/process.rs [hunk 6]
```diff
@@ -946,6 +971,42 @@ impl<'r, 'l, L: SubstateStore> Process<'r, 'l, L> {
...
+        }
+    }
+
+    fn move_lazy_map_into_component(
+        &mut self,
+        unclaimed_lazy_map: UnclaimedLazyMap,
+        mid: Mid,
+        component_address: Address,
+    ) {
+        self.track
```
Comment (iamyulong) [inline radix-engine/src/engine/process.rs:999]: I'd suggest we add some logs to the two move methods. This is extremely useful for debugging through logs (without fancy IDEs).

File: radix-engine/src/engine/process.rs [hunk 7]
```diff
@@ -946,6 +971,42 @@ impl<'r, 'l, L: SubstateStore> Process<'r, 'l, L> {
         Ok(CallMethodOutput { rtn: result?.raw })
     }
 
+    fn move_lazy_map_into_lazy_map(
+        &mut self,
+        unclaimed_lazy_map: UnclaimedLazyMap,
+        mid: Mid,
+        mid_parent: &Mid,
+    ) {
+        let parent_lazy_map = self.unclaimed_lazy_maps.get_mut(&mid_parent).unwrap();
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/process.rs:980]: good point!

[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine/src/engine/process.rs]: There are some structural issues which prevents this refactor from being easy so will push this change to another PR.

**Review by iamyulong — APPROVED** [review]
LGTM. Thanks for dealing with my comments.

**Pull request merged**